frame/ethereum/src/lib.rs

```diff
@@ -150,7 +157,27 @@ where
 		dispatch_info: &DispatchInfoOf<T::RuntimeCall>,
 		len: usize,
 	) -> Option<TransactionValidity> {
-		if let Call::transact { transaction } = self {
+		if let Call::transact { transaction }
+		| Call::transact_with_weight_limit { transaction, .. } = self
+		{
+			// Validate submitted gas limit to weight conversion
+			let gas_limit = match transaction {
+				Transaction::Legacy(t) => t.gas_limit,
+				Transaction::EIP2930(t) => t.gas_limit,
+				Transaction::EIP1559(t) => t.gas_limit,
+			};
+			let without_base_extrinsic_weight = true;
+			let submitted_weight = T::GasWeightMapping::gas_to_weight(
+				gas_limit.unique_saturated_into(),
+				without_base_extrinsic_weight,
+			);
+			if submitted_weight != dispatch_info.weight {
```

Comment:
In your current implementation here, the `submitted_weight` are not submitted but computed. If we do that, it's totally useless to have a new call `transact_with_weight_limit`.
To support sovereign chains (that don't care about the proof size), we can says if the `GasWeightMapping`implementation return zero for the proof_size, we will disable the  dynamic proof size metering.